projects/15_basic_material/main.cpp

```diff
@@ -1038,6 +1250,52 @@ void ProjApp::Render()
         GetGraphicsQueue()->CopyBufferToBuffer(&copyInfo, mCpuModelConstants, mGpuModelConstants, grfx::RESOURCE_STATE_CONSTANT_BUFFER, grfx::RESOURCE_STATE_CONSTANT_BUFFER);
     }
 
+    // Update env draw constants
```

Comment:
In the case of this simple example, it doesn't really matter. Elsewhere on desktop it depends on the API and which memory type and if BAR is exposed or not. If the upload buffer ends up in system memory that's uncached, it will get read across the PCIe bus which would end up detrimental. Uploading to the the GPU guarantees that it will always be read from device memory.